### PR TITLE
[coroutine] Create coroutine body in the correct eval context

### LINKED ISSUE #78290: link error when compiling coroutine code.
See https://godbolt.org/z/5nvhqhsz9 for a minimal testcase.

The following code is compiled with gcc, but clang emits a link error (`in function Foo::CoroutineBody(): <source>:60: undefined reference to CoroutinePromise<void>::CoroutinePromise<Foo&>(Foo&)`) when linking to a binary. We seem to miss a symbol.

```cpp
#include <coroutine>

template <typename Final>
class Gen;

template <typename Final>
class CoroutinePromise final {
 public:
  template <typename... Args>
  explicit CoroutinePromise(Args&&... args) {}

  Gen<Final> get_return_object() { return Gen<Final>(my_handle()); }

  std::coroutine_handle<CoroutinePromise> my_handle() {
    return std::coroutine_handle<CoroutinePromise>::from_promise(*this);
  }
  void unhandled_exception() {}
  void return_void() {}

  template <typename U>
  auto await_transform(Gen<U> gen) {
    struct Awaitable {
      bool await_ready() { return false; }
      std::coroutine_handle<> await_suspend(
          const std::coroutine_handle<> handle_in) {
        return handle_in;
      }
      decltype(auto) await_resume() {}
    };
    return Awaitable();
  }

  std::suspend_always initial_suspend() { return {}; }
  std::suspend_always final_suspend() noexcept { return {}; }
};


template <typename Final>
class Gen final {
 public:
  using final_result_type = Final;
  using promise_type =CoroutinePromise<Final>;
  explicit Gen(const std::coroutine_handle<promise_type> handle)
      : handle_(handle) {}

  std::coroutine_handle<promise_type> handle_;  
};

class Foo  {
 public:
  Gen<void> TestBody2() {
    auto s =  [&]() -> Gen<void> { co_await this->CoroutineBody(); }();
    return s;
  }
 private:
  Gen<void> CoroutineBody();
};

Gen<void> Foo::CoroutineBody() {
  if constexpr (0) // remove it make clang compile.
    co_return;
}

int main() {
  return 0;
}
```

### PR DESCRIPTION
**Pull request by usx95**
Fixes: https://github.com/llvm/llvm-project/issues/78290

See the bug for more context.
```cpp
Gen ACoroutine() {
  if constexpr (0) // remove it make clang compile.
    co_return;
  co_await Gen{};
}
```
We miss symbol of ctor of promise_type if the first coroutine statement happens to be inside the disabled branch of `if constexpr`.

This happens because the promise object is built when we see the first coroutine statement which is present in `ExpressionEvaluationContext::DiscardedStatement` context due to `if constexpr (0)`. This makes clang believe that the promise constructor is only odr-used and not really "used".

The expr evaluation context for the coroutine body should not be related to the context in which the first coroutine statement appears. We override the context to `PotentiallyEvaluated`.

### CODE REVIEW

File: clang/test/SemaCXX/coroutine-promise-ctor.cpp [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+};
+
+Gen CoroutineBody() {
+    if constexpr (0) {
+        co_await Gen{};
+    }
+    co_await Gen{};
+}
+
+int main() { return 0; }
```
Comment (cor3ntin) [inline clang/test/SemaCXX/coroutine-promise-ctor.cpp]: Missing line break

**Review by cor3ntin — COMMENTED** [review]
Can you add a release note for this change (mentioning the issue) ? thanks!

**Comment by usx95** [comment]
Added release notes and fix new line.

File: clang/docs/ReleaseNotes.rst [hunk 2]
```diff
@@ -804,6 +804,9 @@ Bug Fixes in This Version
   Objective-C++ property accesses to not be converted to a function call
   to the getter in the placement-args of new-expressions.
   Fixes (`#65053 <https://github.com/llvm/llvm-project/issues/65053>`_)
+- Fix an issue with missing symbol definitions when the first coroutine
+  statement is part of an disabled branch of ``if constexpr``.
```
Comment (cor3ntin) [inline clang/docs/ReleaseNotes.rst:808]:
```suggestion
  statement appears in a discarded ``if constexpr`` branch.
```

**Review by cor3ntin — APPROVED** [review]
LGTM, thanks!
Please give a chance  to @ChuanqiXu9 to look at it before merging

File: clang/test/SemaCXX/coroutine-promise-ctor.cpp (hunk 1 above)
Comment (hokein) [inline clang/test/SemaCXX/coroutine-promise-ctor.cpp:34]: nit: we do the verification by checking the presence  of CompoundStmt, the main function is not needed.

File: clang/test/SemaCXX/coroutine-promise-ctor.cpp [hunk 3]
```diff
@@ -0,0 +1,34 @@
+// RUN: %clang_cc1 -std=c++20 -ast-dump %s | FileCheck %s
+#include "Inputs/std-coroutine.h"
+
+// Github issue: https://github.com/llvm/llvm-project/issues/78290
```
Comment (hokein) [inline clang/test/SemaCXX/coroutine-promise-ctor.cpp:4]: nit: we usually encode the github issue with `namespace GH78290 {...}`

**Review by hokein — APPROVED** [review]
thanks, this looks good to me.

**Review by ChuanqiXu9 — APPROVED** [review]
LGTM. Thanks.

**Pull request merged**